core/src/main/java/io/parsingdata/metal/expression/value/reference/Offset.java

```diff
@@ -16,32 +16,21 @@
 
 package io.parsingdata.metal.expression.value.reference;
 
-import static io.parsingdata.metal.Util.checkNotNull;
-
 import io.parsingdata.metal.data.Environment;
 import io.parsingdata.metal.data.ParseValue;
 import io.parsingdata.metal.encoding.Encoding;
-import io.parsingdata.metal.expression.value.ConstantFactory;
-import io.parsingdata.metal.expression.value.OptionalValue;
-import io.parsingdata.metal.expression.value.ValueExpression;
-
-public class Offset implements ValueExpression {
-
-    private final String _name;
+import io.parsingdata.metal.expression.value.*;
 
-    public Offset(final String name) {
-        _name = checkNotNull(name, "name");
-    }
+public class Offset extends UnaryValueExpression {
 
-    @Override
-    public OptionalValue eval(final Environment env, final Encoding enc) {
-        final ParseValue ref = env.order.get(_name);
-        return ref != null ? OptionalValue.of(ConstantFactory.createFromNumeric(ref.getOffset(), ref.getEncoding())) : OptionalValue.empty();
-    }
+    public Offset(final ValueExpression op) { super(op); }
 
     @Override
-    public String toString() {
-        return getClass().getSimpleName() + "(" + _name + ")";
+    public OptionalValue eval(final Value op, final Environment env, final Encoding enc) {
+        if (op instanceof ParseValue) {
```

Comment:
Bummer about the `instanceof` and cast code smell, but I see why.

(Maybe discuss this off-line.)